### DISCUSSION TITLE
Horizontal Scaling and AntiForgery key missing exception

### DISCUSSION THREAD

**Discussion by Th3-M4jor** [btcpayserver/btcpayserver#6117] [Technical Support] [answered]
I'm trying to scale BTCPay server horizontally using Kubernetes and whenever I try to log in it returns a 400 with an exception related to the AntiForgery key being missing as the cause.

Is there a way to work around this issue? If it makes a difference, I have no intention of exposing the btcpayserver instances to the public internet.

I saw that there was a 2 year old discussion which said to not run replicas, but maybe this has changed since then?

<details>
<summary>Stacktrace</summary>

```
Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery: An exception was thrown while deserializing the token.
Microsoft.AspNetCore.Antiforgery.AntiforgeryValidationException: The antiforgery token could not be decrypted.
 ---> System.Security.Cryptography.CryptographicException: The key {##UUID KEY REDACTED##} was not found in the key ring. For more information go to https://aka.ms/aspnet/dataprotectionwarning
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.UnprotectCore(Byte[] protectedData, Boolean allowOperationsOnRevokedKeys, UnprotectStatus& status)
   at Microsoft.AspNetCore.DataProtection.KeyManagement.KeyRingBasedDataProtector.Unprotect(Byte[] protectedData)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgeryTokenSerializer.Deserialize(String serializedToken)
   --- End of inner exception stack trace ---
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgeryTokenSerializer.Deserialize(String serializedToken)
   at Microsoft.AspNetCore.Antiforgery.DefaultAntiforgery.GetCookieTokenDoesNotThrow(HttpContext httpContext)
```

</details>

**Comment by pavlenex** [comment]
@NicolasDorier @dennisreimann  do you know anything about this one?

**Reply by dennisreimann** [reply]
I assume one would have to configure or customize the antiforgery validation to use a shared storage for the tokens. The problem you are running into currently is most likely that the instances for generating and validating differ and each use a separate storage/source for the tokens. 

I couldn't quickly find examples, but assume this isn't a very exotic use case, so it should be doable – not sure if this is something for us to solve in core though.

**Reply by Th3-M4jor** [reply]
I've not written any C# in close to a decade at this point, so I could be way off, but would a new configuration option to disable [this line](https://github.com/btcpayserver/btcpayserver/blob/master/BTCPayServer/Hosting/Startup.cs#L87) do the trick?

If that's theoretically the only thing preventing horizontal scaling, I wouldn't mind brushing up on my C# and opening a PR for it.

**Comment by Kukks** [comment, marked as answer]
Do not do this. It will never work without a major in depth refactor. There is a lot of state that is managed or cached in memory, events are dispatched and handled on demand in memory and they modify dB records in a very specific way..

This won't work. Don't do horizontal scaling, you will corrupt your state and run into endless issues. Btcpay is very peformant as a single instance.

This applies to almost all parts of its stack, such as nbx and the ln node
